server/main.ts

```diff
@@ -69,6 +70,18 @@ loadConfig((loadedInitialConfig) => {
     )
   );
 
+  const wss = new WebSocket.Server({ noServer: true });
+  instance.on('upgrade', (req, socket, head) =>
+    wss.handleUpgrade(req, socket, head, (ws) => {
+      const res = new http.ServerResponse(req) as strimziUIResponseType;
+      // add/mark the request as a websocket request
+      req.isWs = true;
+      res.ws = ws;
+      // call the express app as usual
+      expressApp(req, res);
```

Comment:
How does `expressApp` know which endpoint to call? Is it simply that we're listening on `host:port/*` so any websocket request will be upgraded, but then sending messages will fail unless it's on a path that has `isWs` support?

Reply:
Yep, that's right. The code in main.ts (https://github.com/ajborley/strimzi-ui/blob/42-client-side-logging/server/main.ts#L74-L90) watches for websocket upgrade requests, and adds the `isWs` boolean and `ws` websocket to the req and res respectively. So currently, any endpoint that receives an upgrade request will set up a websocket, but only modules that implement the `websocket.onmessage`, etc handlers will actually do anything.

I expect this is not what we want - we should probably add something to the module configuration that can be used to decide whether the websocket upgrade request is permitted or not. What do you think?

Reply:
I wonder if we need a break out session on this.

It seems like we're possibly shoehorning in the `log` module into express, when actually this is a different paradigm. My concern is that by appending `ws` to `req` we lose reference to it outside of the initial request.

Without extending express or using old express middleware - it feels like we need the ability to control "websocket" modules too - maybe, by having the concept of attaching multiple endpoints (https://www.npmjs.com/package/ws#multiple-servers-sharing-a-single-https-server) - we then destroy socket if it's not on a path that's acceptable?

We can then have "express" modules and "ws" modules (with a similar `addModule` for ws that takes in a logger)